### PR TITLE
Tags support for Example sections

### PR DESCRIPTION
**Pull request by elchupanebrej**
This covers #411
- Example tags could be used for test selection
- Example tags could not be used for invoking pytest_bdd_apply_tag hook for now
- Add explicit support for multiple Examples sections
- "Examples: Vertical" replaced with "Examples Transposed:" for more convenient naming
- Examples sections names are parsed (not used now)

### CODE REVIEW

**Comment by felipesetti** [comment]
Hi, What is the status of this issue?

**Comment by KelvinKSPS** [comment]
It is a very expected feature! 
Is there anything pending we can help?

**Comment by elchupanebrej** [comment]
pytest_bdd requires original parser to be rewritten. This PR uses and extends old parser, so seems that this PR won't be accepted till work on new parser would be finished(and has to be re-done after)

I'm in progress of development fork of pytest_bdd which uses cucumber.io gherkin parser, which will support this feature from the box (essential for project which I test), but won't get support of code generation and maybe other features of pytest_bdd on start. I think it could take few weeks to show this feature, because I'm busy on my main project

**Comment by elchupanebrej** [comment]
Also, this feature  aggravate #296 #310

File: pytest_bdd/parser.py [hunk 1]
```diff
@@ -267,12 +288,20 @@ def params(self):
 
     def get_example_params(self):
         """Get example parameter names."""
-        return set(self.examples.example_params + self.feature.examples.example_params)
+        return set(
+            chain(
```
Comment (olegpidsadnyi) [inline pytest_bdd/parser.py:292]: {examples.example_params for examples in chain(...)}?

File: pytest_bdd/parser.py [hunk 2]
```diff
@@ -410,39 +466,24 @@ def add_example_row(self, param, values):
...
-                    raw_value = example[index]
-                    if converters and param in converters:
-                        value = converters[param](raw_value)
-                        if not builtin or value.__class__.__module__ in {"__builtin__", "builtins"}:
-                            example[index] = value
-                params.append(example)
-            return [self.example_params, params]
-        else:
-            return []
+        self.examples = list(zip(*self.example_param_values))
```
Comment (olegpidsadnyi) [inline pytest_bdd/parser.py:476]: that looks weird. why does it assign to self in get_() function instead of return?

File: pytest_bdd/scenario.py [hunk 3]
```diff
@@ -160,14 +161,32 @@ def decorator(*args):
...
+        for params_tags, param_set in scenario.get_params():
+            param_names, param_values = param_set
+            scenario_param_data[tuple(param_names)].append((params_tags, param_values))
+
+        decorator_chain = []
+
+        for param_names in scenario_param_data.keys():
+            parametrize_values = []
+            for param_tags, param_values in scenario_param_data[param_names]:
+                marks = list(map(partial(getattr, pytest.mark), param_tags))
```
Comment (olegpidsadnyi) [inline pytest_bdd/scenario.py:174]: marks = [...]

File: pytest_bdd/scenario.py [hunk 4]
```diff
@@ -160,14 +161,32 @@ def decorator(*args):
...
+                parametrize_values.extend(
+                    [pytest.param(*param_value, marks=marks) for param_value in param_values] if marks else param_values
+                )
+
+            decorator_chain.append(pytest.mark.parametrize(list(param_names), parametrize_values))
+
+        compose = lambda *func: reduce(lambda f, g: lambda x: f(g(x)), func, lambda x: x)
```
Comment (olegpidsadnyi) [inline pytest_bdd/scenario.py:181]: okay.. let's not turn python into haskell

**Review by olegpidsadnyi — CHANGES_REQUESTED** [review]

File: README.rst [hunk 5]
```diff
@@ -496,13 +496,13 @@ pytest-bdd feature file format also supports example tables in different way:
             When I eat <eat> cucumbers
             Then I should have <left> cucumbers
 
-            Examples: Vertical
+            Examples Transposed:
```
Comment (olegpidsadnyi) [inline README.rst:499]: Is it possible to separate tags from transposed examples into separate PRs? It is very hard to follow what's what

**Review by olegpidsadnyi — CHANGES_REQUESTED** [review]

File: README.rst (hunk 5 above)
Comment (elchupanebrej) [reply README.rst:499]: #475

**Comment by elchupanebrej** [comment]
This will be reopened after #475 merge

**Comment by CuriousQA** [comment]
Hey,

I see this is waiting for https://github.com/pytest-dev/pytest-bdd/pull/475, but the other issue is closed now and the branch is deleted. Kindly suggest when would this be available or if there exists any alternatives.

Regards,

**Comment by elchupanebrej** [comment]
@CuriousQA seems, that this feature (and many others attached to it) won't be implemented in this project, because of internal problems that must be unwrapped at first. The biggest one is a rewriting parser of Gherkin files. This activity seems ongoing, so I decided to make a fork of this project, to have the possibility to implement essential features which I need personally. 

That fork supports tags for example sections but doesn't support vertical examples (because they are not a part of the "standard"). You could find it here https://github.com/elchupanebrej/pytest-bdd-ng
